Firestore/core/src/firebase/firestore/remote/grpc_stream.h

```diff
@@ -0,0 +1,206 @@
+ */
+class GrpcStream : public std::enable_shared_from_this<GrpcStream> {
+ public:
+  // Implementation of the stream relies on its memory being managed by
+  // `shared_ptr`.
+  //
+  // The given `grpc_queue` must wrap the same underlying completion queue as
+  // the `call`.
+  static std::shared_ptr<GrpcStream> MakeStream(
+      std::unique_ptr<grpc::ClientContext> context,
+      std::unique_ptr<grpc::GenericClientAsyncReaderWriter> call,
+      GrpcStreamObserver* observer,
+      GrpcCompletionQueue* grpc_queue);
+
+  void Start();
+
+  void Write(grpc::ByteBuffer&& message);
+
+  // Does not produce a notification. Once this method is called, the stream can
+  // no longer be used.
+  //
+  // Can be called on a stream before it opens. It is invalid to finish a stream
+  // more than once.
+  void Finish();
+
+  // Writes the given message and finishes the stream as soon as the write
+  // succeeds. Any non-started writes will be discarded. Neither write nor
+  // finish will notify the observer.
+  //
+  // If the stream hasn't opened yet, `WriteAndFinish` is equivalent to
+  // `Finish` -- the write will be ignored.
+  void WriteAndFinish(grpc::ByteBuffer&& message);
+
+ private:
+  friend class internal::GrpcStreamDelegate;
+
+  GrpcStream(std::unique_ptr<grpc::ClientContext> context,
+             std::unique_ptr<grpc::GenericClientAsyncReaderWriter> call,
+             GrpcStreamObserver* observer,
+             GrpcCompletionQueue* grpc_queue);
+
+  void Read();
+  void BufferedWrite(grpc::ByteBuffer&& message);
+
+  // Called by `GrpcStreamDelegate`.
+  void OnStart();
+  void OnRead(const grpc::ByteBuffer& message);
+  void OnWrite();
+  void OnOperationFailed();
+  void OnFinishedByServer(const grpc::Status& status);
+  void OnFinishedByClient();
+
+  bool SameGeneration() const;
+
+  template <typename Op, typename... Args>
+  Op* MakeOperation(Args... args);
+
+  // Creates and immediately executes an operation; ownership is released.
+  template <typename Op, typename... Args>
+  void Execute(Args... args) {
+    MakeOperation<Op>(args...)->Execute();
+  }
+
+  // The GRPC objects that have to be valid until the last GRPC operation
+  // associated with this call finishes. Note that `grpc::ClientContext` is not
+  // reference-counted.
+  //
+  // Important: `call_` has to be destroyed before `context_`, so declaration
+  // order matters here. Despite the unique pointer, `call_` is actually
+  // a non-owning handle, and the memory it refers to will be released once
+  // `context_` (which is owning) is released.
+  std::unique_ptr<grpc::ClientContext> context_;
+  std::unique_ptr<grpc::GenericClientAsyncReaderWriter> call_;
+
+  GrpcCompletionQueue* grpc_queue_ = nullptr;
+
+  GrpcStreamObserver* observer_ = nullptr;
+  int generation_ = -1;
+  // Buffered writer is created once the stream opens.
+  absl::optional<BufferedWriter> buffered_writer_;
+
+  enum class State {
+    NotStarted,
```

Comment:
There is some duplication of states with Firestore stream here. My feeling was that it's justified by making `GrpcStream` more self-sufficient.